File: built-in plugins dsp/mb-chorus.cpp
/**
 * MB Chorus
 * Category : effect
 * Type     : chorus
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Rich stereo chorus effect
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_CHORUS_H
#define MB_CHORUS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbChorus : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-chorus";
    static constexpr const char* PLUGIN_NAME    = "MB Chorus";
    static constexpr const char* PLUGIN_TYPE    = "chorus";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float rate = 1.0f;  // range [0.1, 10]
    float depth = 0.5f;  // range [0, 1]
    float delay = 7f;  // range [1, 30]
    float spread = 0.7f;  // range [0, 1]
    float mix = 0.5f;  // range [0, 1]
    };

    MbChorus() = default;
    ~MbChorus() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.rate = std::clamp(params.rate, 0.1f, 10f);
        params.depth = std::clamp(params.depth, 0f, 1f);
        params.delay = std::clamp(params.delay, 1f, 30f);
        params.spread = std::clamp(params.spread, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Chorus
        return input;
    }
};

#endif // MB_CHORUS_H

File: built-in plugins dsp/mb-chorus.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbChorusPlugin: PluginDefinition = { id: 'mb-chorus', slug: 'mb-chorus', name: 'MB Chorus', category: 'effect', type: 'chorus', version: '1.0.0', description: 'Rich stereo chorus effect', author: 'Max Booster', parameters: [{ id: 'rate', name: 'Rate', type: 'float', defaultValue: 1.0, minValue: 0.1, maxValue: 10, automatable: true }, { id: 'depth', name: 'Depth', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'delay', name: 'Delay', type: 'float', defaultValue: 7, minValue: 1, maxValue: 30, automatable: true }, { id: 'spread', name: 'Stereo Spread', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }, { id: 'mix', name: 'Dry/Wet Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { rate: 1.0, depth: 0.5, delay: 7, spread: 0.7, mix: 0.5 } };

export default MbChorusPlugin;
